adadamp/_dist.py

```diff
@@ -276,45 +364,84 @@ def run_single_epoch(self, X, y=None, **fit_params):
         """
         dataset = self._get_dataset(X, y=y)
         client = get_client()
+        
+         # copy model
+        module_kwargs = self._get_kwargs_for("module")
+        m = self.module(**module_kwargs)
+        m.to(torch.device(self.device))
+        m.load_state_dict(self.module_.train().state_dict())
 
+        # copy optimizer
+        opt_kwargs = self._get_kwargs_for("optimizer")
+        o = self.optimizer(m.parameters(), **opt_kwargs)
+        o.load_state_dict(self.optimizer_.state_dict())
+        o.zero_grad()
+        
+        assert get_model_grads(m) == 0, "ERROR: Gradients not zero'd at copy time"
+        
         # Send the model/optimizer to workers
-        m = deepcopy(self.module_.train())
         m = client.scatter(m)
-        o = deepcopy(self.optimizer_)
         o = client.scatter(o)
         model_opt = client.submit(lambda x, y: (x, y), m, o)
 
         # Give all data to each worker
         len_dataset = len(dataset)
         dataset = client.scatter(dataset, broadcast=True)
 
+        # Track when we have moved through dataset
         start_data = copy(self._meta["n_data"])
         loss = deepcopy(self.loss_)
+        device = torch.device(self.device)
+        
+        # Run BS items through until hitting every element in dataset
+        prev_weights = 0
         while True:
             model_opt, bs = self.train_step(
                 model_opt,
                 dataset,
                 loss=loss,
                 client=client,
                 len_dataset=len_dataset,
+                device=device,
                 **fit_params,
             )
+
+            
+            # check for weight update
+            model, opt = model_opt.result()
+            
+            new_weights = get_model_weights(model)
+            weights_changed = new_weights != prev_weights
```

Comment:
I'll need to fix this code. Calling `result` isn't necessary, but will require a little finessing.